examples/create_and_manage_prompt_templates_using_structured_output_parser.rb

```diff
@@ -0,0 +1,104 @@
+require "langchain"
+
+# Generate a prompt that directs the LLM to provide a JSON response that adheres to a specific JSON schema.
+json_schema = {
+  type: "object",
+  properties: {
+    name: {
+      type: "string",
+      description: "Persons name"
+    },
+    age: {
+      type: "number",
+      description: "Persons age"
+    },
+    interests: {
+      type: "array",
+      items: {
+        type: "object",
+        properties: {
+          interest: {
+            type: "string",
+            description: "A topic of interest"
+          },
+          levelOfInterest: {
+            type: "number",
+            description: "A value between 0 and 100 of how interested the person is in this interest"
+          }
+        },
+        required: ["interest", "levelOfInterest"],
+        additionalProperties: false
+      },
+      minItems: 1,
+      maxItems: 3,
+      description: "A list of the person's interests"
+    }
+  },
+  required: ["name", "age", "interests"],
+  additionalProperties: false
+}
+parser = Langchain::OutputParsers::StructuredOutputParser.from_json_schema(json_schema)
+prompt = Langchain::Prompt::PromptTemplate.new(template: "Generate details of a fictional character.\n{format_instructions}\nCharacter description: {description}", input_variables: ["description", "format_instructions"])
+prompt.format(description: "Korean chemistry student", format_instructions: parser.get_format_instructions)
```

Comment:
I set this to a variable and then sent to the actual LLM:
```ruby
prompt = prompt.format(description: "Korean chemistry student", format_instructions: parser.get_format_instructions)
openai = Langchain::LLM::OpenAI.new(api_key: ENV['OPENAI_API_KEY'])
response = openai.complete prompt: prompt
parser.parse(response)
#=> Langchain::OutputParsers::OutputParserException (Failed to parse. Text: ")

```json
{
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Person's name"
    },
    "age": {
      "type": "number",
      "description": "Person's age"
    },
    "interests": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "interest": {
            "type": "string",
            "description": "A topic of interest"
          },
          "levelOfInterest": {
            "type": "number",
            "description": "A value between 0 and 100 of how interested the person is in this interest"
          },
          "required": [
            "interest",
            "levelOfInterest"
          ],
          "additionalProperties": false
        },
        "minItems": 1,
        "maxItems": 3,
        "description": "A list of the person's interests"
      },
      "required": [
        "name",
        "age",
        "interests"
      ],
      "additionalProperties": false
    }
  },
  "name": "Jin-Soo",
  "age": 20,
  "interests": [
    {
      "interest": "Chemistry",
      "levelOfInterest": 100
    },
    {
      "interest": "Korean Literature",
      "levelOfInterest": 80
    },
    {
      "interest": "Gardening",
      "levelOfInterest": 50
    }
  ]
}
```"
```
but got an error.

Do you think there's ways to fix it?

It worked with GooglePalm LLM though!

Reply:
I plan to add another feature on top of this `Langchain::OutputParsers::OutputFixingParser` such that if we could send the error back to the LLM questioning them why they got it wrong (again same feature exists in langchainjs) (see example how this could be used below). My experience with openai has shown this to be solid solution. 

Other way i've fixed problem like this is to modify the prompt template to be more explicit about not including the json schema in the response. 

cheers

```
begin
  output = parser.parse(llm_example_response)
rescue => e
  # alternatively, if the StructuredOutputParser fails, try the OutputFixParser
  fix_parser = Langchain::OutputParsers::OutputFixingParser.from_llm(
    llm: Langchain::LLM::OpenAI.new(api_key: ENV["OPENAI_API_KEY"]),
    parser: parser
  )
  output = fix_parser.parse(llm_example_response)
  # {
  #   "name" => "Kim Ji-hyun",
  #   "age" => 22,
  #   "interests" => [
  #     {
  #       "interest" => "Organic Chemistry",
  #       "levelOfInterest" => 85
  #     },
  #     {
  #       "interest" => "Biochemistry",
  #       "levelOfInterest" => 70
  #     },
  #     {
  #       "interest" => "Analytical Chemistry",
  #       "levelOfInterest" => 60
  #     }
  #   ]
  # }
end
```

Reply:
Hmmm... 

> Other way i've fixed problem like this is to modify the prompt template to be more explicit about not including the json schema in the response.

Worth a try.

I can merge this PR and we can continue testing and iterating on it.